### ISSUE TITLE
Money question type

### ISSUE THREAD

**Issue by robmckinnon** [alphagov/smartdown#128]
Are there plans to add a `money` question type?

For a *potential* fee remissions eligibilty smart answers, we require a user to enter a claim fee amount. E.g.

![screen shot 2015-01-08 at 15 30 25](https://cloud.githubusercontent.com/assets/3447/5664845/621078a6-974b-11e4-9dc7-b469fd5b52e4.png)

**Comment by dsingleton** [comment]
No specific plans, but it would happen at some point. Certainly no objection to it being added.

So far we've added new question types to Smartdown as we converted existing smart answers, or built new ones.

There's an implementation of a [`money` question type](https://github.com/alphagov/smart-answers/blob/master/lib/smart_answer/money.rb) in the original smart answers app ([example](https://www.gov.uk/state-pension-topup/y/1947-01-02/male), but we've not needed it in Smartdown yet. 

Adding a new question type would be reasonably straight forward, the most recent example was adding a Postcode question. This PR illustrates the changes required pretty well: https://github.com/alphagov/smartdown/pull/121 (with the caveat of one follow up [bug fix](https://github.com/alphagov/smartdown/pull/123)) and would require some integration into the Smart Answers application.

You might be able to use a [free form text question](https://github.com/alphagov/smartdown/blob/master/spec/parser/element/text_question_spec.rb) as a work around in the short term?

Does that help?

**Comment by robmckinnon** [comment]
For fee remissions, we may want to show different values in later questions depending on the claim fee amount. E.g. savings threshold depends on claim fee like this:
![screen shot 2015-01-09 at 09 46 12](https://cloud.githubusercontent.com/assets/3447/5677914/7ef5ed5e-97e4-11e4-8b58-3302fcf9c29d.png)
etc.

We may want to ask the question like this:
![screen shot 2015-01-09 at 09 54 34](https://cloud.githubusercontent.com/assets/3447/5678011/8e811b9e-97e5-11e4-89df-2f7727be54fb.png)

Seems in Smartdown we can't put conditional blocks nor do variable assignment in questions.

So if we were to implement this, would we need to implement a `money` question type, and a value comparison predicate? E.g. the 

* claim_fee >= '1000' => savings_less_than_3000
* claim_fee > '1000' AND claim_fee <= '1335' => savings_less_than_4000
* claim_fee > '1336' AND claim_fee <= '1665' => savings_less_than_5000
...

**Comment by dsingleton** [comment]
It doesn't support conditional blocks that result in different questions, but you can interpolate inside the text of a question:

```
# Did the mother’s partner work (or will they have worked) for at least 26 weeks between %{earnings_employment_start_date(due_date)} and %{earnings_employment_end_date(due_date)}?

[choice: work_employment_2]
* yes: Yes
* no: No
```

Thats an [example](https://github.com/alphagov/smart-answers/blob/master/lib/smartdown_flows/pay-leave-for-parents/questions/employment_2_more.txt#L3) from the [pay-leave-for-parents](https://github.com/alphagov/smart-answers/blob/master/lib/smartdown_flows/pay-leave-for-parents/) smart answer (which is by far the most complex thing we've implemented in Smartdown, and uses most of it's features, might be worth looking at)

If you wanted actually conditional questions, rather than just varying text, then having multiple next nodes/steps (like `savings_less_than_4000` would be how we'd do it with Smartdown right now.

There are already [comparison predicates](https://github.com/alphagov/smartdown/tree/master/lib/smartdown/model/predicate/comparison), which depend on the `answer` object of a question to be comparable (for example [salary](https://github.com/alphagov/smartdown/blob/master/spec/model/answer/salary_spec.rb#L93-L114) and [date](https://github.com/alphagov/smartdown/blob/master/spec/model/answer/date_spec.rb#L32-L96)). You would need to implement a `money` question type, however - good news - there is already a [`money` answer type](https://github.com/alphagov/smartdown/blob/master/lib/smartdown/model/answer/money.rb). This is used as a return type from [calculations](https://github.com/alphagov/smart-answers/blob/master/lib/smartdown_plugins/pay-leave-for-parents/render_time.rb#L109-L111), so that they can be interpolated into output with proper formatting (which is handled by answer objects, [eg £X.XX for money](https://github.com/alphagov/smartdown/blob/master/spec/model/answer/money_spec.rb)).

**Comment by dsingleton** [comment]
Oh, its worth noting that next node rules rules will pick a rule as soon as it meets a true condition, so...

```
claim_fee >= '1000' => savings_less_than_3000
claim_fee <= '1335' => savings_less_than_4000
```

...is the same as your example (though in some cases we've done the explicit-but-redundant version for readability reasons).

Hope that helps.

**Comment by robmckinnon** [comment]
I've created PRs to add a `money` question type to the smartdown and smart-answers repos:
* smartdown PR: #131 
* smart-answers PR: alphagov/smart-answers#1408

**Comment by robmckinnon** [comment]
Closing, as pull requests to add a money question type have been merged: #131 and alphagov/smart-answers#1408
